tfjs-backend-webgpu/src/kernels/matmul_packed_vec4_webgpu.ts

```diff
@@ -110,6 +110,163 @@ export function makeMatMulPackedVec4Source(workPerThread: number[]): string {
   `;
 }
 
+function getSharedArray1DCodeWgsl() {
+  return `
+  var<workgroup> mm_Asub : array<vec4<f32>, 1024>;`;
+}
+
+function getSharedArray2DCodeWgsl() {
+  return `
+  var<workgroup> mm_Asub : array<array<vec4<f32>, 16>, 64>;
+  var<workgroup> mm_Bsub : array<array<vec4<f32>, 16>, 64>;`;
+}
+
+function getA1DCodeWgsl() {
+  return `
+  let index : u32 = inputRow * TileInner / ColPerThread + inputCol;
+  mm_Asub[index] = mm_readA(globalRow + innerRow, globalColA, globalId);`;
+}
+
+function getA2DCodeWgsl() {
+  return `
+  mm_Asub[inputRow][inputCol] = mm_readA(globalRow + innerRow, globalColA, globalId);`;
+}
+
+function getB1DCodeWgsl() {
+  return `
+  let index : u32 = inputRow * TileBOuter / ColPerThread + inputCol;
+  mm_Bsub[index] = mm_readB(t * TileInner + inputRow, globalCol);`;
+}
+
+function getB2DCodeWgsl() {
+  return `
+  mm_Bsub[inputRow][inputCol] = mm_readB(t * TileInner + inputRow, globalCol);`;
+}
+
+function getCompute1DCodeWgsl() {
+  return `
+  // Compute acc values for a single thread.
+  for (var k : u32 = 0u; k < TileInner / ColPerThread; k = k + 1u) {
+      BCached[0] = mm_Bsub[(k * ColPerThread) * (TileBOuter / ColPerThread) + tileCol];
+      BCached[1] = mm_Bsub[(k * ColPerThread + 1u) * (TileBOuter / ColPerThread) + tileCol];
+      BCached[2] = mm_Bsub[(k * ColPerThread + 2u) * (TileBOuter / ColPerThread) + tileCol];
+      BCached[3] = mm_Bsub[(k * ColPerThread + 3u) * (TileBOuter / ColPerThread) + tileCol];
+
+      for (var i : u32 = 0u; i < RowPerThread; i = i + 1u) {
+          ACached = mm_Asub[(tileRow + i) * (TileInner / ColPerThread) + k];
+
+          acc[i] = BCached[0] * ACached.x + acc[i];
+          acc[i] = BCached[1] * ACached.y + acc[i];
+          acc[i] = BCached[2] * ACached.z + acc[i];
+          acc[i] = BCached[3] * ACached.w + acc[i];
+      }
+  }`;
+}
+
+function getCompute2DCodeWgsl() {
+  return `
+  // Compute acc values for a single thread.
+  for (var k : u32 = 0u; k < TileInner / ColPerThread; k = k + 1u) {
+      BCached[0] = mm_Bsub[k * ColPerThread][tileCol];
+      BCached[1] = mm_Bsub[k * ColPerThread + 1u][tileCol];
+      BCached[2] = mm_Bsub[k * ColPerThread + 2u][tileCol];
+      BCached[3] = mm_Bsub[k * ColPerThread + 3u][tileCol];
+
+      for (var i : u32 = 0u; i < RowPerThread; i = i + 1u) {
+          ACached = mm_Asub[tileRow + i][k];
+          acc[i] = BCached[0] * ACached.x + acc[i];
+          acc[i] = BCached[1] * ACached.y + acc[i];
+          acc[i] = BCached[2] * ACached.z + acc[i];
+          acc[i] = BCached[3] * ACached.w + acc[i];
+      }
+  }`;
+}
+
+function getMainCodeWgsl(getA: string, getB: string, computeAcc: string) {
+  return `
+  let RowPerThread : u32 = 4u;
+  let ColPerThread : u32 = 4u;
+  let TileAOuter : u32 = 64u;
+  let TileBOuter : u32 = 64u;
+  let TileInner : u32 = 64u;
+
+  [[stage(compute), workgroup_size(16, 16, 1)]]
```

Comment:
Can above line put in shader_proprocessor? And we should use program.`workGroupSize` to decide the work group size rather than using a fixed value.

Reply:
done